### ISSUE TITLE
Compression levels question

### ISSUE THREAD

**Issue by r-lyeh** [facebook/zstd#466]
Hello Yann,

During a few tests, I found that lower compression levels may produce smaller results than higher compression levels.
The data is someway controversial though (not real data).

Is this as designed? Will it happen in real data? If so, should I test against all levels before deciding which level should I stick with for every file I want to compress?

```
13893612 -> 11700 ( 0.02 s) L0
13893612 <- 11700 ( 0.02 s)

13893612 -> 11700 ( 0.03 s) L1
13893612 <- 11700 ( 0.02 s)

13893612 -> 7097 ( 0.02 s) L2
13893612 <- 7097 ( 0.02 s)

13893612 -> 4702 ( 0.03 s) L3
13893612 <- 4702 ( 0.01 s)

13893612 -> 4702 ( 0.02 s) L4
13893612 <- 4702 ( 0.01 s)

13893612 -> 19618 ( 0.30 s) L5
13893612 <- 19618 ( 0.02 s)

13893612 -> 43914 ( 0.34 s) L6
13893612 <- 43914 ( 0.02 s)

13893612 -> 17705 ( 0.34 s) L7
13893612 <- 17705 ( 0.01 s)

13893612 -> 17700 ( 0.38 s) L8
13893612 <- 17700 ( 0.01 s)

13893612 -> 17700 ( 0.37 s) L9
13893612 <- 17700 ( 0.01 s)

13893612 -> 4325 ( 0.28 s) L10
13893612 <- 4325 ( 0.01 s)

13893612 -> 4325 ( 0.30 s) L11
13893612 <- 4325 ( 0.01 s)

13893612 -> 2870 ( 0.20 s) L12
13893612 <- 2870 ( 0.01 s)

13893612 -> 2371 ( 0.16 s) L13
13893612 <- 2371 ( 0.01 s)

13893612 -> 2459 ( 0.15 s) L14
13893612 <- 2459 ( 0.01 s)

13893612 -> 1836 ( 0.88 s) L15
13893612 <- 1836 ( 0.02 s)

13893612 -> 1606 ( 0.44 s) L16
13893612 <- 1606 ( 0.01 s)

13893612 -> 2216 ( 0.46 s) L17
13893612 <- 2216 ( 0.01 s)

13893612 -> 2244 ( 0.75 s) L18
13893612 <- 2244 ( 0.01 s)

13893612 -> 2029 ( 0.40 s) L19
13893612 <- 2029 ( 0.01 s)

13893612 -> 2127 ( 0.54 s) L20
13893612 <- 2127 ( 0.02 s)

13893612 -> 2078 ( 0.43 s) L21
13893612 <- 2078 ( 0.01 s)

13893612 -> 1990 ( 0.52 s) L22
13893612 <- 1990 ( 0.02 s)
```

PS: the dataset 

```c++
    // 13 mb dataset
    std::string original( "There's a lady who's sure all that glitters is gold" );
    for (int i = 0; i < 18; ++i) original += original + std::string( i + 1, 32 + i );
```

**Comment by Cyan4973** [comment]
Yes, this can indeed happen, though this particular example is pretty extreme.
There are indeed many possible matches, and selecting the better one can be tricky.

Moving from one level to another involves changing multiple parameters, one of then being `strategy`. Changing `strategy` basically means changing the compression algorithm. Each algorithm makes different bets. In general, they are relatively well ordered, but sometimes, some betting strategies end up being both better and faster. It's not a general rule though.

It seems that the `strategy` of level 3 and 4, named `dFast`, is especially good for this sample. This is unusual, though welcomed. Because of that, moving to level 5, which uses `greedy` strategy is a step back.
Same thing happens at level 16, where strategy `btlazy2` performs better than the more sophisticated `btopt`.

So yes, it can happen. But it's not the "norm". In general, higher levels should perform better.

> should I test against all levels before deciding which level should I stick with for every file I want to compress?

In theory yes, but in practice, this looks excessive.
If you are looking for best possible ratio, level 22 should be your candidate.

By the way, we have recently improved level 20-22 for some specific corner cases (in `dev` branch) where we observed that it was not the best. The improvement was supposed to target small files, but maybe it will help for your use case too.
